Models/PersistentShoppingCart.cs

```diff
@@ -0,0 +1,178 @@
+
+            _products = null;
+        }
+
+        public void AddRange(IEnumerable<ShoppingCartItem> items) {
+            foreach (var item in items) {
+                Add(item.ProductId, item.Quantity, item.AttributeIdsToValues);
+            }
+        }
+
+        public void Clear() {
+            _products = null;
+            _persistentShoppingCartServices.ClearCart();
+            UpdateItems();
+        }
+
+        public void Remove(int productId, IDictionary<int, ProductAttributeValueExtended> attributeIdsToValues = null) {
+            _persistentShoppingCartServices.RemoveItem(productId, attributeIdsToValues);
+            _products = null;
+        }
+
+        public void UpdateItems() {
+            _persistentShoppingCartServices.ConsolidateCart();
+            _products = null;
+        }
+
+        public IEnumerable<ShoppingCartQuantityProduct> GetProducts() {
+            if (_products != null) return _products;
+
+            var ids = Items.Select(x => x.ProductId);
+
+            var productParts =
+                _contentManager.GetMany<ProductPart>(ids, VersionOptions.Published,
+                new QueryHints().ExpandParts<TitlePart, ProductPart, AutoroutePart>()).ToArray();
+
+            var productPartIds = productParts.Select(p => p.Id);
+
+            var shoppingCartQuantities =
+                (from item in Items
+                 where productPartIds.Contains(item.ProductId) && item.Quantity > 0
+                 select new ShoppingCartQuantityProduct(item.Quantity, productParts.First(p => p.Id == item.ProductId), item.AttributeIdsToValues))
+                    .ToList();
+
+            return _products = shoppingCartQuantities
+                .Select(q => _priceService.GetDiscountedPrice(q, shoppingCartQuantities))
+                .Where(q => q.Quantity > 0)
+                .ToList();
+        }
+
+        public decimal Subtotal() {
+            return Math.Round(GetProducts().Sum(pq => Math.Round(pq.Price * pq.Quantity + pq.LinePriceAdjustment, 2)), 2);
+        }
+
+        public TaxAmount Taxes(decimal subTotal = 0) {
+            if (Country == null && ZipCode == null) return null;
+            var taxes = _taxProviders
+                .SelectMany(p => p.GetTaxes())
+                .OrderByDescending(t => t.Priority);
+            var shippingPrice = ShippingOption == null ? 0 : ShippingOption.Price;
+            if (subTotal.Equals(0)) {
+                subTotal = Subtotal();
+            }
+            return (
+                from tax in taxes
+                let name = tax.Name
+                let amount = tax.ComputeTax(GetProducts(), subTotal, shippingPrice, Country, ZipCode)
+                where amount > 0
+                select new TaxAmount { Name = name, Amount = amount }
+                ).FirstOrDefault() ?? new TaxAmount { Amount = 0, Name = null };
+        }
+
+        public decimal Total(decimal subTotal = 0, TaxAmount taxes = null) {
```

Comment:
This too. Maybe a base abstract class for default implementations would be useful here.